### ISSUE TITLE
Failed to start dnsmasq, host name resolution may fail [ macos arm64 ]

### ISSUE THREAD

**Issue by edwinjhlee** [termux/termux-docker#68]
Hi, I want to use termux-docker in macos. However, the dns fail so the hostname resolution fails.

Could you provide some help to fix this problem ? 

```
[!] Failed to start dnsmasq, host name resolution may fail.
```

<img width="606" alt="image" src="https://github.com/user-attachments/assets/bea140df-5dbb-4485-85d8-96026ab62b9a">

## Information about my system

<img width="388" alt="image" src="https://github.com/user-attachments/assets/89e95eaf-cef2-487b-a188-b5e5088db872">

**Comment by Biswa96** [comment]
Does any other Linux based container image work in macos?

**Comment by edwinjhlee** [comment]
> Does any other Linux based container image work in macos?

Yes. It works. 
Actually the termux-docker image works. Only the dnsmasq doesn't works so hostname resolution fails.

**Comment by edwinjhlee** [comment]
I have tried using the privledge mode. Same log and same problem.

<img width="876" alt="image" src="https://github.com/user-attachments/assets/52737ec3-b5f8-461a-85fc-b85333212ae2">

**Comment by edwinjhlee** [comment]
I also try using root-shell following README's instruction.

```
docker run -it --entrypoint /entrypoint_root.sh --privileged termux/termux-docker:aarch64
```

<img width="947" alt="image" src="https://github.com/user-attachments/assets/cbea66d3-877a-4f7d-b53a-42901a8619cb">

**Comment by edwinjhlee** [comment]
Also, I have set the binfmt.

```bash
docker run --rm --privileged aptman/qus -s -- -p aarch64 arm
```

**Comment by mio-19** [comment]
+1

I am using Orbstack for docker

**Comment by robertkirkman** [comment]
I tried for a long time to reproduce this, but unfortunately I have not been able to reproduce this exact specific error.

I believe this means there must be an unknown factor involved, affecting some people but not others, that we need to figure out in order to find a solution that works for everyone.

### Reproduction Attempt Setup

1. To get reproducible freeware MacOS on non-Apple hardware, I downloaded [kholia/OSX-KVM](https://github.com/kholia/OSX-KVM)

2. I configured it to prepare the Nested Virtualization technique that is required to run any Virtualization-based software, including Orbstack and UTM, inside of MacOS Virtual Machine Guests.

> [!IMPORTANT]
> To proceed, the host PC hardware must have an **Intel processor** because AMD processors are not capable of providing Nested Virtualization support within MacOS Virtual Machine Guests.

<details>

<summary>OSX-KVM-enable-nested-virtualization.patch</summary>

```diff
--- a/OpenCore-Boot.sh
+++ b/OpenCore-Boot.sh
@@ -18,7 +18,7 @@
 ###############################################################################
 
 
-MY_OPTIONS="+ssse3,+sse4.2,+popcnt,+avx,+aes,+xsave,+xsaveopt,check"
+MY_OPTIONS="+ssse3,+sse4.2,+popcnt,+avx,+aes,+xsave,+xsaveopt,check,+vmx"
 
 # This script works for Big Sur, Catalina, Mojave, and High Sierra. Tested with
 # macOS 10.15.6, macOS 10.14.6, and macOS 10.13.6.
@@ -33,7 +33,7 @@ OVMF_DIR="."
 
 # shellcheck disable=SC2054
 args=(
-  -enable-kvm -m "$ALLOCATED_RAM" -cpu Penryn,kvm=on,vendor=GenuineIntel,+invtsc,vmware-cpuid-freq=on,"$MY_OPTIONS"
+  -enable-kvm -m "$ALLOCATED_RAM" -cpu host,kvm=on,vendor=GenuineIntel,+invtsc,vmware-cpuid-freq=on,"$MY_OPTIONS"
   -machine q35
   -device qemu-xhci,id=xhci
   -device usb-kbd,bus=xhci.0 -device usb-tablet,bus=xhci.0
@@ -63,6 +63,7 @@ args=(
   # -netdev user,id=net0 -device vmxnet3,netdev=net0,id=net0,mac=52:54:00:c9:18:27  # Note: Use this line for High Sierra
   -monitor stdio
   -device vmware-svga
+  -display sdl
   # -spice port=5900,addr=127.0.0.1,disable-ticketing=on
 )
 
```

</details>

3. I followed the kholia/OSX-KVM installation documentation to completion and chose the MacOS 13 Ventura guest.

4. Inside of the guest, I downloaded and installed [OrbStack version 1.5.1](https://docs.orbstack.dev/downgrade#v1-5-1)

5. I opened a new Terminal window inside the guest and ran:
  - `docker run --rm --privileged aptman/qus -s -- -p aarch64 arm` 
  - `docker run -it --privileged termux/termux-docker:aarch64` 

6. I ran `curl google.com` to confirm that `dnsmasq` is functioning.

![Image](https://github.com/user-attachments/assets/41ad1f14-623e-42c4-8ba4-9ea2bba7d41e)

Maybe posting this can help narrow down possible factors that could be required to reproduce the issue.

**Comment by edwinjhlee** [comment]
Thanks for the reply. I will try orbstack and podman desktop later to see if I can provide more information.

BTW, is there any command that I can collect more logs or any other information about the failure related to dnsmasq ?

**Comment by robertkirkman** [comment]
Yes there is, that could be very helpful if you can use these commands, and tell me what you see happen.

```bash
docker run -it --privileged --entrypoint /entrypoint_root.sh termux/termux-docker:aarch64
/system/bin/mksh -T /dev/ptmx -c "/system/bin/dnsmasq -u root -g root --pid-file /dnsmasq.pid"
echo $?
```
That is the command that is not working for you. Maybe if you unsuppress its output by running it this way, we can see an error message with more information.

**Comment by edwinjhlee** [comment]
<img width="1039" alt="Image" src="https://github.com/user-attachments/assets/9dfafc64-780a-45dd-9cb1-521591893308" />

**Comment by robertkirkman** [comment]
Very good, thank you. This gives a lot of useful information about what is happening.

As you can see, this is a situation in which **even though you tried both the `--privileged` argument and the `--security-opt` argument**, as directed by the current README,

**you still experienced a `personality`-related error of some kind**.

This seems like it might be similar, but not identical, to a situation that I have also experienced, not while using a Mac, but while using a [BeagleBone Black device](https://www.beagleboard.org/boards/beaglebone-black).

When I ran termux-docker on the BeagleBone Black, not just the `dnsmasq` program, but the entire container, failed to work, even while trying the `--privileged` argument and the `--security-opt` argument. I was also seeing errors referencing "`personality`".

However, I am able to successfully work around my problem, and run termux-docker directly on the BeagleBone Black, by patching and compiling a custom build of the Docker Daemon source code that has modified `personality` system call permissions. After I did that, all of the errors disappeared and running `docker run -it termux/termux-docker:arm` works normally while the modifed Docker Daemon is running.

I have included this patch and instructions for installing it into my PR to the termux-docker repository, but here is a direct link to the modifed README I created describing it. https://github.com/robertkirkman/termux-docker/blob/deblobbify/README.md

I do not know if the same exact patch would help you on Mac, or not, without being able to reproduce the problem on my own Mac virtual machine. As I have currently written it, there are also no directions for how to install it onto a Mac, only Debian. 

I believe it might be either impossible or very difficult to install the patch into the "Docker Desktop" or "OrbStack" Docker Daemon implementations for Mac, because "Docker Desktop" and "OrbStack" are both closed-source.

However, **[UTM](https://github.com/utmapp/UTM) is a fully open source Virtualization frontend for Mac, and it _is_ possible to run Docker Daemon inside of Debian inside of UTM**.

If it's not too inconvenient for you, would it be possible for you to install UTM, and install Debian or another full GNU/Linux guest inside of UTM, and install Docker inside that guest, and then check whether that is a possible way to work around or avoid this error happening to you?

If the error does not happen anymore after you do that, that would imply the problem is influenced by something inside of the Virtualized Linux kernel or Docker Daemon of OrbStack or Docker Desktop specifically,

but if the error is still happening after that, despite changing the entire Virtualized Linux kernel and Docker Daemon to completely standard builds within a full GNU/Linux guest, then maybe the patch I created for the BeagleBone Black device has a chance to help.

**Comment by edwinjhlee** [comment]
> If it's not too inconvenient for you, would it be possible for you to install UTM, and install Debian or another full GNU/Linux guest inside of UTM, and install Docker inside that guest, and then check whether that is a possible way to work around or avoid this error happening to you?

 I'll investigate using UTM later and ask my Mac-using colleagues to test the image to see if the issue is reproducible. 

Thank you for your help.

**Comment by edwinjhlee** [comment]
<img width="1061" alt="Image" src="https://github.com/user-attachments/assets/b4c34909-75e1-4dfe-98a8-4ca217935b78" />

I run a debian on UTM VM. And I am encoutering the same issue of dnsmasq.

**Comment by edwinjhlee** [comment]
Is the latest aarch64 image broken ? I found /system/bin/sh can not be executed. Maybe it is of x64/aarch64 related problem.

<img width="1102" alt="Image" src="https://github.com/user-attachments/assets/1064ce41-10c0-4248-9f34-2ebfc7a034a3" />

**Comment by robertkirkman** [comment]
Could you send the result of the `uname -m` command on your device, outside of Docker?

I also experienced `Exec format error` when attempting to run termux-docker on one of my other devices (not the BeagleBone Black), 

a device where `uname -m` prints "`aarch64`", and in that case, I was able to determine that my device has a **[Cortex-A34](https://developer.arm.com/Processors/Cortex-A34)-like** `aarch64` CPU.

Cortex-A34-like `aarch64` CPUs (unlike Cortex-A72-like `aarch64` CPUs) are not capable of natively executing 32-bit ARM binaries, in any operating system, only 64-bit ARM binaries, because they lack the hardware feature to support 32-bit binaries.

My PR,

- #72 

contains some changes to the internal components of termux-docker, that **will fix that problem** by ensuring each of the four architecture variants of the termux-docker image is generated using exactly the correct binaries for that architecture. You can read more about it in my PR summary there.

My PR is not merged yet partly because it is waiting on some dependencies in another repository to progress, so unfortunately you cannot get it automatically yet. If you are very enthusiastic about testing it, you could clone my PR and use its `generate.sh` to generate a local Docker image, 

but in general I would recommend waiting for the PR to get future updates, because one of its goals is to **switch to binaries entirely compiled through CI**, so that **the trustworthiness of the binaries in the image  is rigorous**. That part is not ready yet.

**Comment by edwinjhlee** [comment]
In macos:

```
⚜ uname -m
arm64
```


In UTM debian:

```
~ # uname -m
aarch64

```

**Comment by robertkirkman** [comment]
Yes, you have an Apple Silicon Mac, and when I check now, I can see that the `dnsmasq` binary in the current `termux/termux-docker:aarch64` image is also 32-bit ARM. 

```
$ file dnsmasq 
dnsmasq: ELF 32-bit LSB shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, interpreter /system/bin/linker, BuildID[md5/uuid]=4901b7a480227525be19f030ef591e01, stripped
```

I am not sure why you didn't get the `Exec format error` for that one too, but I wrongly assumed that you had an Intel Mac at the start, so I'm sorry for wasting time, I should have asked that much sooner.

I don't have an Apple Silicon Mac to test, but its processor is probably behaving in a similar way to how my Cortex-A34-like device is behaving, so I will mark this issue as one of those my PR can solve once it's ready.

**Comment by edwinjhlee** [comment]
I'm assuming we're talking about Mac ARM...  I have tested it on a MacBook Intel and it worked.
    
You're too kind; my issue should come with more information, I am sorry that I wasted a lot of your time.
    
I'm using Termux on macOS just to test x-cmd, a project I'm working on. Some of our users in China are eager to run x-cmd on their phones. However, we've recently lowered its support priority, so this issue isn't urgent for me. 
    
I really appreciate your PR and the work you've done above. Unfortunately, I don't know much about Termux or its development, so I can't help with that.


Just in case you have interest: https://www.x-cmd.com/

**Comment by thesayyn** [comment]
Just want to give gratitude to @robertkirkman your PR fixes this problem.
